src/libopensc/card-itacns.c

```diff
@@ -544,6 +566,19 @@ itacns_card_ctl(sc_card_t *card, unsigned long cmd, void *ptr)
 	return SC_ERROR_NOT_SUPPORTED;
 }
 
+static int
+itacns_get_challenge(sc_card_t *card, u8 *rnd, size_t len)
+{
+	if (card->type == SC_CARD_TYPE_ITACNS_CNS_IDEMIA_2021) {
+		/* Hardware surely supports this, but the proprietary library simulates
+		 * this in software, so we don't know how to achieve this in hw yet
```

Comment:
I'm curious, did you actually test the RNG with the default_ops->get_challenge for the new card?

Reply:
Indeed:

<details><summary>log for `--generate-random 64`</summary>

```
P:2551629; T:0x140228730681344 16:16:08.444 [opensc-pkcs11] pkcs11-global.c:626:C_GetSlotInfo: C_GetSlotInfo() card detect rv 0x0
P:2551629; T:0x140228730681344 16:16:08.444 [opensc-pkcs11] pkcs11-global.c:643:C_GetSlotInfo: C_GetSlotInfo() flags 0x7
P:2551629; T:0x140228730681344 16:16:08.444 [opensc-pkcs11] pkcs11-global.c:647:C_GetSlotInfo: C_GetSlotInfo(0x8) = CKR_OK
Using slot 2 with a present token (0x8)
P:2551629; T:0x140228730681344 16:16:08.444 [opensc-pkcs11] pkcs11-session.c:58:C_OpenSession: C_OpenSession(0x8)
P:2551629; T:0x140228730681344 16:16:08.445 [opensc-pkcs11] slot.c:470:slot_get_token: Slot(id=0x8): get token
P:2551629; T:0x140228730681344 16:16:08.445 [opensc-pkcs11] slot.c:488:slot_get_token: Slot-get-token returns OK
P:2551629; T:0x140228730681344 16:16:08.445 [opensc-pkcs11] pkcs11-session.c:94:C_OpenSession: C_OpenSession handle: 0x55fa103fcc40
P:2551629; T:0x140228730681344 16:16:08.445 [opensc-pkcs11] pkcs11-session.c:97:C_OpenSession: C_OpenSession() = CKR_OK
P:2551629; T:0x140228730681344 16:16:08.445 [opensc-pkcs11] card.c:919:sc_get_challenge: called
P:2551629; T:0x140228730681344 16:16:08.446 [opensc-pkcs11] card.c:473:sc_lock: called
P:2551629; T:0x140228730681344 16:16:08.446 [opensc-pkcs11] reader-pcsc.c:688:pcsc_lock: called
P:2551629; T:0x140228730681344 16:16:08.446 [opensc-pkcs11] card.c:513:sc_lock: returning with: 0 (Success)
P:2551629; T:0x140228730681344 16:16:08.446 [opensc-pkcs11] apdu.c:550:sc_transmit_apdu: called
P:2551629; T:0x140228730681344 16:16:08.446 [opensc-pkcs11] card.c:473:sc_lock: called
P:2551629; T:0x140228730681344 16:16:08.446 [opensc-pkcs11] card.c:513:sc_lock: returning with: 0 (Success)
P:2551629; T:0x140228730681344 16:16:08.447 [opensc-pkcs11] apdu.c:515:sc_transmit: called
P:2551629; T:0x140228730681344 16:16:08.447 [opensc-pkcs11] apdu.c:363:sc_single_transmit: called
P:2551629; T:0x140228730681344 16:16:08.447 [opensc-pkcs11] apdu.c:367:sc_single_transmit: CLA:0, INS:84, P1:0, P2:0, data(0) (nil)
P:2551629; T:0x140228730681344 16:16:08.447 [opensc-pkcs11] reader-pcsc.c:325:pcsc_transmit: reader 'Alcor Micro AU9560 00 00'
P:2551629; T:0x140228730681344 16:16:08.447 [opensc-pkcs11] reader-pcsc.c:326:pcsc_transmit: 
Outgoing APDU (5 bytes):
00 84 00 00 40 ....@

P:2551629; T:0x140228730681344 16:16:08.447 [opensc-pkcs11] reader-pcsc.c:244:pcsc_internal_transmit: called
P:2551629; T:0x140228730681344 16:16:08.454 [opensc-pkcs11] reader-pcsc.c:335:pcsc_transmit: 
Incoming APDU (2 bytes):
67 00 g.

P:2551629; T:0x140228730681344 16:16:08.454 [opensc-pkcs11] apdu.c:382:sc_single_transmit: returning with: 0 (Success)
P:2551629; T:0x140228730681344 16:16:08.455 [opensc-pkcs11] apdu.c:539:sc_transmit: returning with: 0 (Success)
P:2551629; T:0x140228730681344 16:16:08.455 [opensc-pkcs11] card.c:523:sc_unlock: called
P:2551629; T:0x140228730681344 16:16:08.455 [opensc-pkcs11] iso7816.c:128:iso7816_check_sw: Wrong length
P:2551629; T:0x140228730681344 16:16:08.455 [opensc-pkcs11] iso7816.c:686:iso7816_get_challenge: GET CHALLENGE failed: -1206 (Wrong length)
P:2551629; T:0x140228730681344 16:16:08.455 [opensc-pkcs11] card.c:523:sc_unlock: called
P:2551629; T:0x140228730681344 16:16:08.455 [opensc-pkcs11] reader-pcsc.c:740:pcsc_unlock: called
P:2551629; T:0x140228730681344 16:16:08.456 [opensc-pkcs11] card.c:934:sc_get_challenge: returning with: -1206 (Wrong length)
P:2551629; T:0x140228730681344 16:16:08.456 [opensc-pkcs11] misc.c:71:sc_to_cryptoki_error_common: libopensc return value: -1206 (Wrong length)
P:2551629; T:0x140228730681344 16:16:08.456 [opensc-pkcs11] pkcs11-object.c:1319:C_GenerateRandom: C_GenerateRandom() = CKR_DATA_LEN_RANGE
error: Could not generate random bytes
Aborting.
```
</details>

Reply:
If you want I can pass you the pcapng for the proprietary driver, but that's also just using a software implementation, there's no get_challenge call at all there.

Reply:
I just wondered why you want to restrict OpenSC's functionality to that of the proprietary driver...

The response code 6700 indicates that you've requested too many bytes. Try to request 8 bytes, for example, i.e. `00 84 00 00 08`

Reply:
Oh, you're right... I thought that 64 bits were just  a mininum value (as previous card was able to fetch up to 256 IIRC), so wondering if we should just support bigger requests by repeating them to fill the request?

Reply:
This should automatically be done by the upper layers of OpenSC. Just fill in as much random data as your card supports and return the number of bytes that you've filled in.

Reply:
Yeah I was thinking the same while doing it, but I wanted an upstream suggestion on that. IIRC it wasn't able to fill up the buffer that way while I wanted to add the same logic to higher levels, but wasn't sure if was a wanted generic behavior.